src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestTPFatTPFServerWithTranslation.java

```diff
@@ -0,0 +1,236 @@
+ */
+public class ExecOpRequestTPFatTPFServerWithTranslation extends ExecOpGenericTriplePatternRequestWithTPF<TPFServer>
+{
+	public ExecOpRequestTPFatTPFServerWithTranslation( final TriplePatternRequest req, final TPFServer fm ) {
+		super( req, fm );
+	}
+
+	@Override
+	protected TPFResponse performPageRequest( final TPFRequest req,
+	                                          final FederationAccessManager fedAccessMgr )
+			throws FederationAccessException
+	{
+		final SPARQLGraphPattern reqTranslation = fm.getVocabularyMapping().translateTriplePattern(req.getQueryPattern());
+		final List<Triple> resList = new ArrayList<>();
+		if(reqTranslation instanceof TriplePattern) {
+			resList.addAll(handleTriplePattern((TriplePattern) reqTranslation, fedAccessMgr));
+		}
+		
+		else if(reqTranslation instanceof SPARQLUnionPattern) {
+			resList.addAll(handleUnionPattern(((SPARQLUnionPattern) reqTranslation), fedAccessMgr));
+		}
+		
+		else if(reqTranslation instanceof SPARQLGroupPattern) {
+			resList.addAll(handleGroupPattern(((SPARQLGroupPattern) reqTranslation), fedAccessMgr));
+		}
+		
+		else if(reqTranslation instanceof BGP) {
+			resList.addAll(handleBGP((BGP) reqTranslation, fedAccessMgr));
+		}
+		
+		else {
+			throw new FederationAccessException(reqTranslation.toString(), req, fm);
+		}
+		
+		return new TPFResponseImpl(resList, resList, null, fm, req, null);
+	}
+	
+	protected Set<Triple> handleTriplePattern(final TriplePattern tp, final FederationAccessManager fedAccessMgr) throws FederationAccessException{
+		final TPFRequest newReq = new TPFRequestImpl(tp);
+		final TPFResponse res = FederationAccessUtils.performRequest(fedAccessMgr, newReq, fm);
+		final Set<Triple> tripleTranslation = new HashSet<>();
+		for(final Triple i : res.getPayload()) {
+			tripleTranslation.addAll(translateResultTriple(i, newReq));
+		}
+		return tripleTranslation;
+	}
+	
+	protected Set<Triple> handleUnionPattern(final SPARQLUnionPattern up, final FederationAccessManager fedAccessMgr) throws FederationAccessException{
+		final Set<Triple> unionTranslation = new HashSet<>();
+		for(final SPARQLGraphPattern i : up.getSubPatterns()) {
+			if (i instanceof TriplePattern) {
+				unionTranslation.addAll(handleTriplePattern(((TriplePattern) i), fedAccessMgr));
+			} else if (i instanceof SPARQLUnionPattern) {
+				unionTranslation.addAll(handleUnionPattern(((SPARQLUnionPattern) i), fedAccessMgr));
+			} else if (i instanceof SPARQLGroupPattern) {
+				unionTranslation.addAll(handleGroupPattern(((SPARQLGroupPattern) i), fedAccessMgr));
+			} else if (i instanceof BGP) {
+				unionTranslation.addAll(handleBGP(((BGP) i), fedAccessMgr));
+			} else {
+				throw new FederationAccessException(i.toString(), req, fm);
+			}
+		}
+		return unionTranslation;
+	}
+	
+	protected Set<Triple> handleGroupPattern(final SPARQLGroupPattern gp, final FederationAccessManager fedAccessMgr) throws FederationAccessException{
+		Set<Triple> groupTranslation = null;
```

Comment:
Same here, I think that we need to perform a join between the results of evaluating each part of the Group Pattern, instead of performing a union.